packages/eslint-plugin/src/rules/no-unnecessary-type-parameters.ts

```diff
@@ -0,0 +1,346 @@
+            !identifierCounts ||
+            identifierCounts.general > 2 ||
+            identifierCounts.return > 2 ||
+            (identifierCounts.general === 2 && identifierCounts.return) ||
+            (identifierCounts.return === 2 && identifierCounts.general)
+          ) {
+            continue;
+          }
+
+          context.report({
+            data: {
+              name: typeParameter.name.text,
+            },
+            node: parserServices.tsNodeToESTreeNodeMap.get(typeParameter),
+            messageId: 'sole',
+          });
+        }
+      },
+    };
+  },
+});
+
+interface IdentifierCounts {
+  general: number;
+  return: number;
+}
+
+type IdentifierArea = keyof IdentifierCounts;
+
+/**
+ * Count uses of type parameters in inferred return types.
+ * We need to resolve and analyze the inferred return type of a function
+ * to see whether it contains additional references to the type parameters.
+ * For classes, we need to do this for all their methods.
+ */
+function countTypeParameterUsage(
+  checker: ts.TypeChecker,
+  node: NodeWithTypeParameters,
+): Map<ts.Identifier, IdentifierCounts> {
+  const counts = new Map<ts.Identifier, IdentifierCounts>();
+
+  if (ts.isClassLike(node)) {
+    for (const typeParameter of node.typeParameters) {
+      collectTypeParameterUsageCounts(checker, typeParameter, counts);
+    }
+    for (const member of node.members) {
+      collectTypeParameterUsageCounts(checker, member, counts);
+    }
+  } else {
+    collectTypeParameterUsageCounts(checker, node, counts);
+  }
+
+  return counts;
+}
+
+/**
+ * Fills in a counter of the number of times each type parameter appears in
+ * the given type by recursively descending through type references.
+ * This is essentially a limited subset of the scope manager, but for types.
+ */
+function collectTypeParameterUsageCounts(
+  checker: ts.TypeChecker,
+  node: ts.Node,
+  identifierCounts: Map<ts.Identifier, IdentifierCounts>,
+): void {
+  const visitedSymbolLists = new Set<ts.Symbol[]>();
+  const type = checker.getTypeAtLocation(node);
+  const typeCounts = new Map<ts.Type, number>();
+  const visitedConstraints = new Set<ts.TypeNode>();
+  let functionLikeType = false;
+  let visitedDefault = false;
+
+  if (
+    ts.isCallSignatureDeclaration(node) ||
+    ts.isConstructorDeclaration(node)
+  ) {
+    functionLikeType = true;
+    visitSignature(checker.getSignatureFromDeclaration(node), 'general');
+  }
+
+  if (!functionLikeType) {
+    visitType(type, 'general', false);
+  }
+
+  function visitType(
+    type: ts.Type | undefined,
+    area: IdentifierArea,
+    asRepeatedType: boolean,
+  ): void {
+    // Seeing the same type > (threshold=3 ** 2) times indicates a likely
+    // recursive type, like `T = { [P in keyof T]: T }`.
+    // If it's not recursive, then heck, we've seen it enough times that any
+    // referenced types have been counted enough to qualify as used.
+    if (!type || incrementTypeCount(type) > 9) {
```

Comment:
> `incrementTypeCount(type) > 9`

For the life of me, I can't figure out an isolated unit test case that fails without this check. But when running it on the repository:

```plaintext


RangeError: Maximum call stack size exceeded
Occurred while linting /Users/josh/repos/typescript-eslint/packages/utils/src/eslint-utils/RuleCreator.ts:59
Rule: "@typescript-eslint/no-unnecessary-type-parameters"
    at getReducedApparentType (/Users/josh/repos/typescript-eslint/node_modules/typescript/lib/typescript.js:58024:36)
    at getIndexInfosOfType (/Users/josh/repos/typescript-eslint/node_modules/typescript/lib/typescript.js:58385:44)
    at getIndexInfoOfType (/Users/josh/repos/typescript-eslint/node_modules/typescript/lib/typescript.js:58388:28)
    at getIndexTypeOfType (/Users/josh/repos/typescript-eslint/node_modules/typescript/lib/typescript.js:58392:20)
    at Object.getIndexTypeOfType (/Users/josh/repos/typescript-eslint/node_modules/typescript/lib/typescript.js:46409:43)
    at TypeObject.getNumberIndexType (/Users/josh/repos/typescript-eslint/node_modules/typescript/lib/typescript.js:145924:31)
    at visitType (/Users/josh/repos/typescript-eslint/packages/eslint-plugin/dist/rules/no-unnecessary-type-parameters.js:185:28)
    at visitTypesList (/Users/josh/repos/typescript-eslint/packages/eslint-plugin/dist/rules/no-unnecessary-type-parameters.js:241:13)
    at visitType (/Users/josh/repos/typescript-eslint/packages/eslint-plugin/dist/rules/no-unnecessary-type-parameters.js:148:13)
    at visitType (/Users/josh/repos/typescript-eslint/packages/eslint-plugin/dist/rules/no-unnecessary-type-parameters.js:186:13)
```

Personally, I'd be (barely) satisfied filing a followup issue given how tricky this rule has gotten. But it'd be great to figure something out now...